Fixes grabbing of legend when cursor is above the legend during scrolling of plot

yusufjimoh opened the pull request:
### Description of work
The default behaviour in Matplotlib is that the legend gets grabbed 
when the cursor is over it during a plot scroll event.
To fix this issue, I set the legend's draggability to false during scrolling 
and re-enable it once the scrolling ends.

Fixes: #37041 

### To test:

### To test:
1. Create a plot
2. scroll the plot to zoom in/out, 
3. scroll when the cursor is above the legend
4. the cursor doesn't grab the legend

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

robertapplin committed:
Remove qt connection from IETPresenter to DataReductionTab

robertapplin committed:
Remove validate from slots section

robertapplin committed:
Move algorithm runner into DataReductionTab

robertapplin committed:
Execute ISISEnergyTransfer algorithm using algorithm runner

robertapplin committed:
Handle algorithm complete and error for ISISIET algorithm

robertapplin committed:
Make possible to execute queue of algorithms

robertapplin committed:
IETPresenter is no longer a QObject

robertapplin committed:
Store last executed algorithm in AlgorithmRunner

robertapplin committed:
Tidy up changes in IET Presenter

robertapplin committed:
Rename AlgorithmRunner to QtAlgorithmRunner because it uses Qt signals

robertapplin committed:
Move AlgorithmRunner to MantidQtWidgets common

robertapplin committed:
Remove unused includes and functions

robertapplin committed:
Update failing tests after introducing AlgorithmRunner

robertapplin committed:
Create basics of FittingPresenter

robertapplin committed:
Move FittingModel and FitPropertyBrowser to FittingPresenter

robertapplin committed:
Subscribe FittingPresenter to FitPropertyBrowser

robertapplin committed:
Make updateFitStatus a private method

robertapplin committed:
Pass AlgorithmRunner into FittingPresenter

robertapplin committed:
Move call of fit algorithm to FittingPresenter

robertapplin committed:
Create IFitPlotModel class and change ownership to FittingModel

robertapplin committed:
Combine addOutput methods into one method

robertapplin committed:
Combine cleanFailedRun methods

robertapplin committed:
Remove unused updateSingleFitOutput function

robertapplin committed:
Tidy up changes for running single fits

robertapplin committed:
Move class methods to correct private or public sections

robertapplin committed:
Get Inelastic tests to build again

robertapplin committed:
Pass FitData raw pointer directly to FitPlotModel

robertapplin committed:
Update FitPlotPresenterTest

robertapplin committed:
Pass FitOutput directly into FitPlotModel

robertapplin committed:
Create FittingPresenterTest file

robertapplin committed:
Add tests for FittingPresenter

robertapplin committed:
Fix non-virtual destructor and other warnings

rosswhitfield committed:
Add new disableReplaceWSButton Property flag. Disable button creation on the WANDPowderReduction algorithm.

rosswhitfield committed:
Add release notes

adriazalvarez committed:
create MVP tabs passing the unique_ptr of the model

adriazalvarez committed:
Initialize elwin presenter passing model pointer

adriazalvarez committed:
Initialize moments presenter passing model pointer

adriazalvarez committed:
Initialize Iqt presenter passing model pointer

adriazalvarez committed:
Initialize sqw presenter passing model pointer

adriazalvarez committed:
Initialize symmetrise presenter passing model pointer

adriazalvarez committed:
Initialize tab widgets explicitly

adriazalvarez committed:
Convert to const some members of sqw and moments models

Commit:
Add note about missing wavelength tooltips

This is a known issue, but low priority for the SANS group right now.

RE #35085

jhaigh0 committed:
added section to packaging documentation

jhaigh0 committed:
review suggestions

Co-authored-by: Gui Maciel Pereira <80104863+GuiMacielPereira@users.noreply.github.com>

jhaigh0 committed:
set doc tests to false on dev docs only

jhaigh0 committed:
Update dev-docs/source/Packaging.rst

Co-authored-by: Gui Maciel Pereira <80104863+GuiMacielPereira@users.noreply.github.com>

rosswhitfield committed:
Add log compression option to AlignAndFocusPowder

SilkeSchomann committed:
Temporary pin libcxx in env

adriazalvarez committed:
Fix crash by selecting rows and add select all button

adriazalvarez committed:
Add release notes

adriazalvarez committed:
Add missing function in test

SilkeSchomann committed:
Disable PythonChannel test

rbauststfc committed:
Add new version of SURF IDF

rbauststfc committed:
Copy GUI values that we want to keep from the main state object rather than assigning a reference to the original object

rbauststfc committed:
Set the row state output name to None if nothing has been set to prevent mistakenly using a name from another row

rbauststfc committed:
Add release notes for issue 37356

adriazalvarez committed:
Add virtual destructors to interface

Commit:
Clarify file numbers produced

RE #37373

Commit:
Fix workspace counts

RE #37373

GuiMacielPereira committed:
Fixed handling of errorbar containers

GuiMacielPereira committed:
Fixed commit

adriazalvarez committed:
Add catch for runtime error

adriazalvarez committed:
Update reduction docs with missing info

GuiMacielPereira committed:
Added release note

adriazalvarez committed:
Fix typo in elwin instructions

cailafinn committed:
Add missing commas

Improve the grammar.

RE #37373

robertapplin committed:
Fix error when switching from Normal fitting to TF Asymmetry fitting

robertapplin committed:
Add release note

robertapplin committed:
Update FDA testing instructions

adriazalvarez committed:
Remove tie if it catches runtime error

robertapplin committed:
Update HIFI test instructions

robertapplin committed:
Update PSI instructions to give location of dolly 1529

GuiMacielPereira committed:
Added unit test

adriazalvarez committed:
Remove tie with correct index

adriazalvarez committed:
Add release notes

peterfpeterson committed:
Remove jemalloc from osx

It is not included in the package dependencies. This also sorts the
dependencies so diff shows more meaningful information when comparing
the platforms.

peterfpeterson committed:
Match version pinning from developer requirements

v5.2.1 has an issue for some tests, v5.3.0 doesn't want to work
on rhel9

peterfpeterson committed:
Change platform tags from unix to osx or linux

peterfpeterson committed:
Update conda/recipes/mantid/meta.yaml

Co-authored-by: thomashampson <thomas.hampson@stfc.ac.uk>

yusufjimoh committed:
fix: fixes scrolling when cursor is above legend with grab the legend - issue 37041

yusufjimoh committed:
fix: fixes scrolling when cursor is above legend with grab the legend - issue 37041

GuiMacielPereira commented on qt/applications/workbench/workbench/plotting/figureinteraction.py:
```diff
@@ -136,6 +136,15 @@ def on_scroll(self, event):
+        """
+        legend = event.inaxes.axes.get_legend()
+        legend_set_draggable(legend, False)
+
```
This is a good start. I would suggest having an if statement to check whether the legend is already set to draggable or not. Because if the legend is not draggable already, we do not want to accidentally set it to true.

```suggestion
    if legend.get_draggable() and legend.contains(event):
        legend_set_draggable(legend, False)
        legend_set_draggable(legend, True)
```

GuiMacielPereira commented on qt/applications/workbench/workbench/plotting/figureinteraction.py:
```diff
@@ -144,6 +153,9 @@ def on_scroll(self, event):
         event.canvas.draw()
 
+        """Re-enable the legend's draggability after the scroll event finishes"""
+        legend_set_draggable(legend, True)
```
I would delete these lines and put this inside the if statement I suggested above

GuiMacielPereira commented on qt/applications/workbench/workbench/plotting/figureinteraction.py:
```diff
@@ -136,6 +136,15 @@ def on_scroll(self, event):
+        Disable the legend's draggability when the scroll event is
+        enabled to prevent the legend from unintentionally moving,
+        which is the default behavior in Matplotlib.
+        """
```
These kinds of triple quotes comments should be used only at the beginning of functions, and they're called docstrings:  https://peps.python.org/pep-0257/. 
Instead you can make a standard comment:
```suggestion
# Correcting for picking legend with scroll event, which is default in Matplotlib
```

GuiMacielPereira reviewed (changes requested):
Somehow looks like you have accidentally removed some files that are not related to this issue.  You can probably fix this by reverting this commit on your local machine, implementing the changes I have suggested and then pushing the changes again to this remote branch :)

yusufjimoh committed:
Revert "fix: fixes scrolling when cursor is above legend with grab the legend - issue 37041"

This reverts commit 5f81d6b0242e763a926bb7eb86576ed9c228fa8b.

yusufjimoh committed:
Revert "fix: fixes scrolling when cursor is above legend with grab the legend - issue 37041"

This reverts commit 5f81d6b0242e763a926bb7eb86576ed9c228fa8b.

yusufjimoh committed:
Add function to correct for legend being picked

Added function that checks if scroll event was inside legend
and prevents it bring picked by setting its dragability

- If legend is not set to draggable, then nothing happens
- If legend is set to draggable, then we set draggable to False
and then to True imediatly afterwards. This weird fix avoids bug
in Matplotlib where legend is picked by default

yusufjimoh committed:
Add function to correct for legend being picked

Added function that checks if scroll event was inside legend
and prevents it bring picked by setting its dragability

- If legend is not set to draggable, then nothing happens
- If legend is set to draggable, then we set draggable to False
and then to True imediatly afterwards. This weird fix avoids bug
in Matplotlib where legend is picked by default

GuiMacielPereira reviewed (changes requested):
Needs a unit test and a release note, you can find guidance for release notes in https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html#release-notes-guide

Commit:
Add unit test

`This unit test checks that legend.set_draggable() was called twice,
which is the implementation we found to get around matplotlib default behaviour.

Co-authored-by: Gui Maciel Pereira <gui.maciel-pereira@stfc.ac.uk>

Commit:
Add unit test

`This unit test checks that legend.set_draggable() was called twice,
which is the implementation we found to get around matplotlib default behaviour.

Co-authored-by: Gui Maciel Pereira <gui.maciel-pereira@stfc.ac.uk>

Commit:
Add release note

Commit:
Add release note

GuiMacielPereira commented on qt/applications/workbench/workbench/plotting/test/test_figureinteraction.py:
```diff
@@ -731,7 +742,7 @@ def _create_mock_fig_manager_to_accept_right_click(self):
         return fig_manager
 
-    def _create_mock_fig_manager_to_accept_middle_click(self):
+    def _create_mock_fig_manager_to_accept_middle_click(self) -> object:
```
```suggestion
    def _create_mock_fig_manager_to_accept_middle_click(self):
```
I would put it back as it was, since specifying that `object` is returned is not very helpful, as everything in Python is an `object`

GuiMacielPereira commented on qt/applications/workbench/workbench/plotting/figureinteraction.py:
```diff
@@ -144,6 +147,13 @@ def on_scroll(self, event):
         event.canvas.draw()
 
+    def correct_for_scroll_event_on_legend(self, event):
+        # Corrects bug in Matplotlib where legend is picked up by scroll event
```
```suggestion
        # Corrects default behaviour in Matplotlib where legend is picked up by scroll event
```

GuiMacielPereira commented on qt/applications/workbench/workbench/plotting/figureinteraction.py:
```diff
@@ -136,6 +136,9 @@ def on_scroll(self, event):
         ):
             return
+
+        self.correct_for_scroll_event_on_legend(event)
```
```suggestion
        self._correct_for_scroll_event_on_legend(event)
```
Put underscore at beginning of name to signify that function is only used in this file

GuiMacielPereira commented on qt/applications/workbench/workbench/plotting/figureinteraction.py:
```diff
@@ -144,6 +147,13 @@ def on_scroll(self, event):
         self.redraw_annotations()
         event.canvas.draw()
 
+    def correct_for_scroll_event_on_legend(self, event):
```
```suggestion
    def _correct_for_scroll_event_on_legend(self, event):
```

GuiMacielPereira commented on qt/applications/workbench/workbench/plotting/test/test_figureinteraction.py:
```diff
@@ -755,12 +766,17 @@ def _create_mock_middle_click(self):
         return mouse_event
 
-    def _create_mock_double_left_click(self):
+    def _create_mock_double_left_click(self) -> object:
```
```suggestion
    def _create_mock_double_left_click(self):
```
Same as above

GuiMacielPereira reviewed (changes requested):
Some tiny changes to the code are required.
Other than that I checked:

- Fix works as expected, `scroll_event` does not pick up legend anymore 
- Fix works on both MantidAxes and normal Matplotlib axes
- If legend draggable is `False`, `scroll_event` does not change it
- Unit test passing, also checked that it fails if a change is made in the code

Happy to approve once small suggestions are added to the code

yusufjimoh committed:
Apply suggestions from code review

Co-authored-by: Gui Maciel Pereira <80104863+GuiMacielPereira@users.noreply.github.com>

yusufjimoh committed:
Apply suggestions from code review

Co-authored-by: Gui Maciel Pereira <80104863+GuiMacielPereira@users.noreply.github.com>

Commit:
Merge branch '37041-scrolling-when-cursor-is-above-legend-will-grab-the-legend' of https://github.com/mantidproject/mantid into 37041-scrolling-when-cursor-is-above-legend-will-grab-the-legend

Commit:
Try to fix merge conflicts

I tried to rebase main but was getting some errors.
Resolved merge conflicts by doing a merge from the main branch instead.

GuiMacielPereira commented:
The mess of commits was caused by trying to fix the merge conflicts through a merge of the main branch. Opted to open a new PR and cherry-pick only the actual good commits.

Closing this PR as it is no longer needed.
